File: OBS_for_CRLB_ISAC/Fig1_convergence/propose_SCA.m
clc
clear

Nth=4;
Ntv=4;
Nt=Nth*Ntv;

Nrh=5;
Nrv=4;
Nr=Nrh*Nrv;

channel_number=50;
K=4;
M=2;

rng(0)
theta=-pi/3+2*pi/3*rand(M,1);
phi=-pi/3+2*pi/3*rand(M,1);

num_sensing_streams=Nt;
tolerance=1e-5;

delta_all=-7:0.05:4.8;

Per_all=zeros(2,length(delta_all));
Con=cell(3,1);

for channel=2
for de=1:3
lin=[0.05,0.1,0.15];
disp([channel,de])
delta_s=1;
delta_c=lin(de);
%10^(delta_all(de));


Pt=db2pow(10-30); %dBm
noise_c=db2pow(0-30); %dBm
noise_s=db2pow(0-30); %dBm



rng(channel)

alpha=0.1*(1+0.2*randn(M,1)).*exp(1j*2*pi*rand(M,1));


L=30;

% theta=[-pi/6, pi/6, pi/10];
% phi=[-pi/3, pi/6, pi/10];
 

kappa=2*L/noise_s;

H=1/sqrt(2)*(randn(Nt,K)+1j*randn(Nt,K));


tic
[A,dAtheta,dAphi]=construct_steer_matrix_and_derivative_steer_matrix(theta,phi,Nth,Ntv);
[B,dBtheta,dBphi]=construct_steer_matrix_and_derivative_steer_matrix(theta,phi,Nrh,Nrv);

U=diag(alpha);

Wc=randn(Nt,K)+1j*randn(Nt,K);
% Ws=delta_s*repmat(eigvector,1,num_sensing_streams);
 Ws=randn(Nt,num_sensing_streams)+1j*randn(Nt,num_sensing_streams);

% Ws=initial_Ws(L,noise_s,Nt,num_sensing_streams,A,dAtheta,dAphi,B,dBtheta,dBphi,U);
W=[Wc,Ws];


W=W*sqrt(Pt/trace(W*W'));

FIM=calculateFIM(L,noise_s,W,A,dAtheta,dAphi,B,dBtheta,dBphi,U);
W_last=W;


for count=1:2000

T_k=sum(square_abs(H'*W(:,1:K)),2)+noise_c*ones(K,1);
alpha_k=T_k./(T_k-square_abs(diag(H'*W(:,1:K))))-1;
beta_k=sqrt(1+alpha_k).*diag(H'*W(:,1:K))./T_k;
Sigma1=diag(sqrt(1+alpha_k).*beta_k);
Sigma2=diag(square_abs(beta_k));

CRBM=inv(FIM);

% PHI=CRBM*CRBM;
Q=construct_matrixQ(L,noise_s,CRBM*CRBM,A,dAtheta,dAphi,B,dBtheta,dBphi,U);
C1=[delta_c*H*Sigma1,zeros(Nt,num_sensing_streams)];
C2=-0.5*delta_s*(Q+Q')+delta_c*H*Sigma2*H';

%dom eigvalue
mu=abs(eigs(C2,1,'LM'));
C2=mu*eye(Nt)-C2;
for iter=1
W=C1+C2*W;

% gradient
% mu=5e-5;
% W= W + mu *(C1-C2*W );

W=W*sqrt(Pt/trace(W*W'));
end

FIM=calculateFIM(L,noise_s,W,A,dAtheta,dAphi,B,dBtheta,dBphi,U);

 T_k=sum(square_abs(H'*W),2)+noise_c*ones(K,1);

 rate_p=log(T_k./(T_k-square_abs(diag(H'*W))));
 % CRBM=delta_s*trace(inv(FIM));
 obj=delta_c*sum(log(T_k./(T_k-square_abs(diag(H'*W)))))-delta_s*trace(inv(FIM));
 Con{de}=[Con{de};[sum(log(T_k./(T_k-square_abs(diag(H'*W))))),-trace(inv(FIM)),obj]];

 if norm(W-W_last)<tolerance
     break
 else
     W_last=W;
 end

end


Rx=W*W';
CRB=trace(inv(FIM));
SR=sum(log(T_k./(T_k-square_abs(diag(H'*W)))));

Per(:,de)=[CRB;SR];
toc
figure(1)
hold on
plot(Con{de}(:,1))
xlabel('iteration')
ylabel('Sum Rate (nat/s/Hz)')
figure(2)
hold on
plot(Con{de}(:,2))
xlabel('iteration')
ylabel('trace of inverse of FIM')
figure(3)
hold on
plot(Con{de}(:,3))
xlabel('iteration')
ylabel('total objective value')

end

% Per_all=Per+Per_all;
end

% save data
% plot(Per_all(1,:)/channel_number,Per_all(2,:)/channel_number)
% plot(Per(1,:),Per(2,:))
% 

save data_convergence Con


function Ws=initial_Ws(L,noise_s,Nt,num_sensing_streams,A,dAtheta,dAphi,B,dBtheta,dBphi,U)

Ws=randn(Nt,num_sensing_streams)+1j*randn(Nt,num_sensing_streams);


for iter=1:5
    FIM=calculateFIM(L,noise_s,Ws,A,dAtheta,dAphi,B,dBtheta,dBphi,U);
    CRBM=inv(FIM);
    Q=construct_matrixQ(L,noise_s,CRBM*CRBM,A,dAtheta,dAphi,B,dBtheta,dBphi,U);
Ws=(Q+Q')*Ws;
Ws=Ws*sqrt(1/trace(Ws*Ws'));
end


end

File: python-port/Fig1_convergence/propose_SCA.py
import numpy as np
from scipy.io import savemat
from scipy.sparse.linalg import eigs
import matplotlib.pyplot as plt
from time import perf_counter
import os
import sys
import argparse

sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from utils.steering_matrix import construct_steer_matrix_and_derivative_steer_matrix
from utils.calculate_fim import calculateFIM
from utils.construct_matrixQ import construct_matrixQ
from utils.square_abs import square_abs
from utils.simulation_config import SimulationConfig

# Not used
def initial_Ws(L, noise_s, Nt, num_sensing_streams, A, dAtheta, dAphi, B, dBtheta, dBphi, U):
    """Initialize sensing beamforming matrix"""
    Ws = np.random.randn(Nt, num_sensing_streams) + 1j * np.random.randn(Nt, num_sensing_streams)

    for _ in range(5):
        FIM = calculateFIM(L, noise_s, Ws, A, dAtheta, dAphi, B, dBtheta, dBphi, U)
        CRBM = np.linalg.inv(FIM)
        Q = construct_matrixQ(L, noise_s, CRBM @ CRBM, A, dAtheta, dAphi, B, dBtheta, dBphi, U)
        Ws = (Q + Q.conj().T) @ Ws
        Ws = Ws * np.sqrt(1 / np.trace(Ws @ Ws.conj().T))

    return Ws

def main():
    # Parse command line arguments
    parser = argparse.ArgumentParser(
        description='OBS-for-CRLB-ISAC Figure 1 Convergence Analysis',
        parents=[SimulationConfig.create_argument_parser()],
        conflict_handler='resolve'
    )

    # Add figure-specific arguments
    parser.add_argument('--save-plots', action='store_true',
                        help='Save plots as files instead of displaying')
    parser.add_argument('--output-dir', type=str, default='.',
                        help='Directory to save output files')

    args = parser.parse_args()

    # Initialize configuration from arguments
    config = SimulationConfig.from_args(args)
    config = config.get_scenario_config('fig1')

    # Print configuration
    print(config)
    print()

    # Get configuration parameters
    Nth = config.Nth
    Ntv = config.Ntv
    Nt = config.Nt

    Nrh = config.Nrh
    Nrv = config.Nrv
    # NOTE:
    # Nr is not used
    # Nr = config.Nr

    # NOTE:
    # channel_number is not used
    # channel_number = config.channel_number
    K = config.K
    # NOTE:
    # No need to set M here
    # Because it's used internally in config.generate_target_angles
    # M = config.M

    num_sensing_streams = config.num_sensing_streams
    tolerance = config.tolerance
    max_iterations = config.max_iterations

    delta_s = config.delta_s
    lin = config.convergence_lin_values  # [0.05, 0.1, 0.15]

    Pt = config.Pt
    noise_c = config.noise_c
    noise_s = config.noise_s

    L = config.L
    # NOTE:
    # kappa is not used
    # kappa = config.kappa

    # Initialize random number generator
    rng = np.random.default_rng(config.random_seed)

    # Generate target angles using config method
    theta, phi = config.generate_target_angles(rng)

    delta_all = config.get_delta_range()
    # NOTE:
    # Per_all is not used
    Per_all = np.zeros((2, len(delta_all)))
    Con = [[], [], []]

    for channel in [2]:
        for de in range(3):
            delta_c = lin[de]

            # Generate alpha using config method
            rng = np.random.default_rng(channel)
            alpha = config.generate_alpha(rng)

            H = 1/np.sqrt(2) * (rng.standard_normal((Nt, K)) + 1j * rng.standard_normal((Nt, K)))

            # MATLAB uses tic
            start_time = perf_counter()

            A, dAtheta, dAphi = construct_steer_matrix_and_derivative_steer_matrix(theta, phi, Nth, Ntv)
            B, dBtheta, dBphi = construct_steer_matrix_and_derivative_steer_matrix(theta, phi, Nrh, Nrv)

            U = np.diag(alpha)

            Wc = rng.standard_normal((Nt, K)) + 1j * rng.standard_normal((Nt, K))
            Ws = rng.standard_normal((Nt, num_sensing_streams)) + 1j * rng.standard_normal((Nt, num_sensing_streams))

            # TODO: Why hstack is used?
            W = np.hstack([Wc, Ws])
            W = W * np.sqrt(Pt / np.trace(W @ W.conj().T))

            FIM = calculateFIM(L, noise_s, W, A, dAtheta, dAphi, B, dBtheta, dBphi, U)
            W_last = W.copy()

            for count in range(max_iterations):
                T_k = np.sum(square_abs(H.conj().T @ W[:, :K]), axis=1) + noise_c * np.ones(K)
                alpha_k = T_k / (T_k - square_abs(np.diag(H.conj().T @ W[:, :K]))) - 1
                beta_k = np.sqrt(1 + alpha_k) * np.diag(H.conj().T @ W[:, :K]) / T_k

                Sigma1 = np.diag(np.sqrt(1 + alpha_k) * beta_k)
                Sigma2 = np.diag(square_abs(beta_k))

                CRBM = np.linalg.inv(FIM)
                Q = construct_matrixQ(L, noise_s, CRBM @ CRBM, A, dAtheta, dAphi, B, dBtheta, dBphi, U)

                C1 = np.hstack([delta_c * H @ Sigma1, np.zeros((Nt, num_sensing_streams), dtype=complex)])
                C2 = -0.5 * delta_s * (Q + Q.conj().T) + delta_c * H @ Sigma2 @ H.conj().T

                # Use largest magnitude eigenvalue to match MATLAB's eigs(C2,1,'LM')
                # eigenvals = linalg.eigvals(C2)
                # mu = np.abs(eigenvals[np.argmax(np.abs(eigenvals))])
                mu = np.abs(eigs(C2, k=1, which='LM', return_eigenvectors=False)[0])
                C2 = mu * np.eye(Nt) - C2

                W = C1 + C2 @ W
                W = W * np.sqrt(Pt / np.trace(W @ W.conj().T))

                FIM = calculateFIM(L, noise_s, W, A, dAtheta, dAphi, B, dBtheta, dBphi, U)
                print(f'FIM: {FIM}')

                # TODO: Why axis=1
                T_k = np.sum(square_abs(H.conj().T @ W), axis=1) + noise_c * np.ones(K)
                # TODO: Why the dot of the first T_k. is omitted
                rate_p = np.log(T_k / (T_k - square_abs(np.diag(H.conj().T @ W))))
                # TODO: Why the dot of the first T_k. is omitted
                obj = delta_c * np.sum(np.log(T_k / (T_k - square_abs(np.diag(H.conj().T @ W))))) - \
                    delta_s * np.trace(np.linalg.inv(FIM))

                # Append data to Con[de] to match MATLAB behavior
                Con[de].append([
                    # TODO: Why the dot of the first T_k. is omitted
                    np.sum(np.log(T_k / (T_k - square_abs(np.diag(H.conj().T @ W))))),
                    -np.trace(np.linalg.inv(FIM)),
                    obj
                ])

                # Check convergence
                norm_diff = np.linalg.norm(W - W_last)

                if norm_diff < tolerance:
                    break
                else:
                    # Only update W_last when NOT converged, matching MATLAB logic
                    W_last = W.copy()

                    # TODO: Why does this line exist
            Con[de] = np.array(Con[de])

            Rx = W @ W.conj().T
            CRB = np.trace(np.linalg.inv(FIM))
            # TODO: Why the dot of the first T_k. is omitted
            SR = np.sum(np.log(T_k / (T_k - square_abs(np.diag(H.conj().T @ W)))))

            print(f"Configuration: lin={lin[de]}")
            print(f"Converged after {len(Con[de])} iterations")
            print(f"Final objective value: {Con[de][-1, 2]:.6f}")
            print(f"Sum Rate: {SR:.6f}")
            print(f"CRB: {CRB:.6f}")
            print(f"Time elapsed: {perf_counter() - start_time:.6f} seconds")
            print()

            # Plotting
            plt.figure(1)
            plt.plot(Con[de][:, 0], label=f'lin={lin[de]}')
            plt.xlabel('iteration')
            plt.ylabel('Sum Rate (nat/s/Hz)')
            plt.legend(loc='lower right')

            plt.figure(2)
            plt.plot(Con[de][:, 1], label=f'lin={lin[de]}')
            plt.xlabel('iteration')
            plt.ylabel('trace of inverse of FIM')
            plt.legend(loc='lower right')

            plt.figure(3)
            plt.plot(Con[de][:, 2], label=f'lin={lin[de]}')
            plt.xlabel('iteration')
            plt.ylabel('total objective value')
            plt.legend(loc='lower right')

    # Save results - Con should be saved directly to match MATLAB format
    # Convert to object array format that MATLAB uses for cell arrays
    # Create a (3,1) object array to match MATLAB's cell array structure
    Con_matlab = np.empty((3, 1), dtype=object)
    for i in range(3):
        Con_matlab[i, 0] = Con[i]

    # Save to output directory
    output_file = os.path.join(args.output_dir, 'data_convergence.mat')
    os.makedirs(args.output_dir, exist_ok=True)

    savemat(output_file, {
        'Con': Con_matlab  # Use (3,1) object array to match MATLAB cell array
    })
    print(f"Results saved to: {output_file}")

    if args.save_plots:
        # Save plots as files
        plot_names = [
            'fig1_sum_rate_vs_iteration.png',
            'fig1_inverse_fim_trace_vs_iteration.png', 
            'fig1_objective_value_vs_iteration.png'
        ]

        for fig_num, plot_name in enumerate(plot_names, 1):
            plot_file = os.path.join(args.output_dir, plot_name)
            plt.figure(fig_num)
            plt.tight_layout()
            plt.savefig(plot_file, dpi=300, bbox_inches='tight')
            print(f"Plot {fig_num} saved to: {plot_file}")
        plt.close('all')
    else:
        plt.show()

if __name__ == "__main__":
    main()
